### ISSUE TITLE
Feature Request: Add compatibility for Firefox Mobile (Android)

### ISSUE THREAD

**Issue by myano** [einaregilsson/Redirector#29]
Could you please add compatibility for Firefox Mobile for Android please?

**Comment by einaregilsson** [comment]
Sorry, that's probably never gonna happen, unless someone else submits a pull request. I don't use Android or Redirector, and this project is really just in maintenance mode. I'll leave this issue open if someone else wants to tackle it.

**Comment by myano** [comment]
@einaregilsson Okay, thanks for the heads up, completely understandable. I appreciate the honesty. :-)

If anybody finds this from googling or any other means, please, please, please, comment on this or reach out to me if you figure out how to do this on Android *within* Firefox. Thanks. :-)

**Comment by Gitoffthelawn** [comment]
Einar is very willing to have people contribute to the project.  I made UI improvements a while back, and it was excellent (and easy) to work with Einar.

**Comment by einaregilsson** [comment]
No one should work on this without first looking at https://github.com/einaregilsson/Redirector/issues/30 to see where Redirector is going.

**Comment by KrasnayaPloshchad** [comment]
Current version is available on Firefox for Android, but I don’t know how to open the setting page.

**Comment by einaregilsson** [comment]
Does Firefox on Android not have toolbar buttons? If it does the Redirector button should be available. Alternatively you can go to resource://redirector-at-einaregilsson-dot-com/redirector.html and bookmark that.

**Comment by KrasnayaPloshchad** [comment]
Firefox for Android doesn’t support toolbar buttons, and this addon seems not useful even if I opened redirector.html

**Comment by einaregilsson** [comment]
I've just published 3.0.6 which adds a preferences button, which should work on Android. It's awaiting review right now, should be out in the next couple of days.

**Comment by KrasnayaPloshchad** [comment]
On Firefox for Android, this addon noticed “Redirector does not work in Private Mode” even if open the setting page in normal mode.

**Comment by Gitoffthelawn** [comment]
@einaregilsson It looks like Redirector does work in Private Mode in the desktop version of Firefox, but does not work in the Android version of Firefox.  Is my understanding correct?

**Comment by einaregilsson** [comment]
Redirector doesn't work in private mode, I just updated it to show a message on the page instead of showing some angular code. I'm not sure what's going on with Android, I'll try to look into it, but it probably won't be for at least a few weeks.

**Comment by KrasnayaPloshchad** [comment]
@einaregilsson I think you can try Bluestacks to run Firefox for Andoid if you want to know what's going on with Android.

**Comment by Gitoffthelawn** [comment]
@einaregilsson wrote "Redirector doesn't work in private mode, I just updated it to show a message on the page instead of showing some angular code. I'm not sure what's going on with Android, I'll try to look into it, but it probably won't be for at least a few weeks."

First, when you say "on the page", to which page are you referring?

Also, I just gave it a try in a Private Window in Firefox 44.0.2 (desktop), and it seems to work.  I entered a URL in the urlbar that has a Redirector rule, and it redirected as expected.

**Comment by einaregilsson** [comment]
That is really weird, because the Redirects page definitely doesn't work in Private mode, that's why I thought the entire addon would be disabled! I need to look into it, and the Android stuff, but I don't have time right now.

**Comment by Gitoffthelawn** [comment]
@einaregilsson No problem.  I just tried loading the Redirector configuration page in Private mode, and I got "Redirector doesn't work in private mode" message.  I guess that's the one you just added.  I don't know what it did before you added that.  Outside of that, Redirector seems to work in Private mode, but I haven't tested it extensively.

No rush from me at all.  Look at it when it's fun for you!  :-)

**Comment by KrasnayaPloshchad** [comment]
> @einaregilsson I think you can try Bluestacks to run Firefox for Andoid if you want to know what's going on with Android.

You can also try to run Android-x86 on VMware Player or VBox to run Firefox for Mobile, which has x86 version in their FTP site.

**Comment by einaregilsson** [comment]
Sorry, still don't have time for this. I'm basically just waiting until WebExtensions is mature enough in Firefox that I can just use the normal Chrome extension in Firefox as well, then the extension should work on Android as well. But anyone is free to investigate this issue further if they're interested.

**Comment by KrasnayaPloshchad** [comment]
> Does Firefox on Android not have toolbar buttons? If it does the Redirector button should be available. Alternatively you can go to resource://redirector-at-einaregilsson-dot-com/redirector.html and bookmark that.

This page does not work since 38.0.

**Comment by KrasnayaPloshchad** [comment]
I found there is a way to find & open the configuration page, input _moz-extension://<your extension ID>/redirector.html_ in location bar and press Enter would open it. This ID is available at about:config page if you installed this extension in your phone, by inputing __extensions.webextensions.uuids__ option.

**Comment by einaregilsson** [comment]
Thanks for the workaround :)

**Comment by KrasnayaPloshchad** [comment]
Firefox Mobile will bring WebEx support in 56.0.
https://www.mozilla.org/en-US/firefox/android/56.0beta/releasenotes/

**Comment by einaregilsson** [comment]
I don't really use Android, but if someone wants to contribute they're more than welcome to, I'm not really working on this extension anymore, but will release new versions if someone contributes a patch.

**Comment by shirish93** [comment]
Hey @einaregilsson , I've been playing with making firefox addons a bit, so I can give it a shot!

So I want to get a sense of what the issue here is: Is it "Redirector doesn't work on android", or is it "such and such API doesn't work on android", or something else?

I tried to get it working on my android, and the settings page didn't pop up, so I have a solid sense where to start from! Also, since I have looked at the mdn docs for webextensions to the point my eyes were bleeding, I also have a general sense of the limitations or possible issues in porting this to a working version in android (the android API is a lot limited than general webextension API).

Is there anything else I should know before I begin/have any suggestions so that I may avoid wild goose chases?

**Comment by shirish93** [comment]
Hello,

I have a tentative fix for this issue for upto Firefox 56, and then a different one for Firefox 57+ (android). Would a pull request be worth the effort?

**Comment by einaregilsson** [comment]
Hi @shirish93 , sorry for the late reply. What is the current Firefox release? Is it worth having special cases to support older releases? I would definitely look at a pull request and merge it in if it was working well.

**Comment by KrasnayaPloshchad** [comment]
The current Firefox release is 56.0, and the 57.0 release will coming in the next weeek.

**Comment by einaregilsson** [comment]
Ok, I'd recommend just targeting 57.

**Comment by Kusresa** [comment]
@einaregilsson

I am wondering what is the status of Redirector for Android? It seems to install and I can access settings page but I don't know if the redirect functionality is working properly and/or optimized for mobile. @shirish93 are you still planning to improve compatibility for Android?

**Comment by einaregilsson** [comment]
I honestly have no idea. It's not something that I'm explicitly supporting, but it might work. I'd welcome any pull request to give first class support for it, but don't have time (or interest frankly) to work on it myself.

**Comment by gkrishnaks** [comment]
@KrasnayaPloshchad, @Kusresa  I know it's an old thread.  Redirector works fine in Firefox for Android in 59.  But yes, the settings page UI the overlay form is very hard to edit in phone - I suggest writing your rules in Desktop Firefox and export it as JSON and import it in Firefox Android.

**Comment by KrasnayaPloshchad** [comment]
@gkrishnaks I think the settings page UI can be reconstruct using responsive web design.

**Comment by gkrishnaks** [comment]
@KrasnayaPloshchad, yes, I'm aware of RWD.  I am just saying... when it works just fine in desktop version, we can easily write rules with phsycial keyboard (detailed regex for include pattern, exclude pattern, base64 , mainframe subframe and other advanced settings stuff) there in a Desktop /laptop and export json and import in android firefox version in few minutes.
